### PR TITLE
Don't recurse in typed array constructor

### PR DESCRIPTION
**Pull request by anba**

### CODE REVIEW

**Comment by anba** [comment]
The recursion happens in this case:
```js
class MyInt8 extends Int8Array {
  constructor(...args) {
    print("MyInt8 called with:", args);
    super(...args);
  }
}
// Prints:
// MyInt8 called with: 1,2,3
// MyInt8 called with: 3
new MyInt8([1, 2, 3]);
```

File: spec.html [hunk 1]
```diff
@@ -30984,16 +30984,19 @@
...
           1. If _thisArg_ was supplied, let _t_ be _thisArg_; else let _t_ be *undefined*.
-          1. Return TypedArrayFrom(_C_, _source_, _f_, _t_).
+          1. Return ? TypedArrayFrom(*undefined*, _C_, _source_, _f_, _t_).
         </emu-alg>
 
         <!-- es6num="22.2.2.1.1" -->
         <emu-clause id="sec-typedarrayfrom" aoid="TypedArrayFrom">
-          <h1>Runtime Semantics: TypedArrayFrom( _constructor_, _items_, _mapfn_, _thisArg_ )</h1>
-          <p>When the TypedArrayFrom abstract operation is called with arguments _constructor_, _items_, _mapfn_, and _thisArg_, the following steps are taken:</p>
+          <h1>Runtime Semantics: TypedArrayFrom( _O_, _constructor_, _items_, _mapfn_, _thisArg_ )</h1>
```
Comment (zenparsing) [inline spec.html]: The signature of this AO is a bit strange now, since you either pass the first argument or the second.  Is there some refactoring that needs to be done to clean this up?

Comment (anba) [inline spec.html]: I've just reintroduced the _strangeness_. :stuck_out_tongue: 

(A similar signature was present in ES6 rev24-rev30, except the _O_ and _constructor_ parameters are switched now.)

Let me think about this a bit, maybe it can be improved...

File: spec.html [hunk 2]
```diff
@@ -31010,7 +31013,19 @@
                   1. Let _nextValue_ be ? IteratorValue(_next_).
                   1. Append _nextValue_ to the end of the List _values_.
               1. Let _len_ be the number of elements in _values_.
-              1. Let _targetObj_ be ? TypedArrayCreate(_C_, &laquo;_len_&raquo;).
+              1. If _O_ is not *undefined*, then
+                1. Let _constructorName_ be the String value of _O_'s [[TypedArrayName]] internal slot.
```
Comment (bterlson) [inline spec.html]: Probably worth factoring this out into another TypedArrayCreate-like AO.

**Comment by bterlson** [comment]
The various ways of creating TAs are getting hard to follow. Perhaps a high level overview of the available AOs and how/when to use them is in order?

**Comment by anba** [comment]
Added two new commits:
 1. 31bfc5a73147d830763b88866074b6ca78e4efa7
  - Moves the common ArrayBuffer allocation to a new abstract operation (AllocateTypedArrayBuffer).
  - Adds descriptions to TypedArrayCreate and TypedArraySpeciesCreate (text copied more or less copied from ArraySpeciesCreate)

 1. 07b8b991b462dcfabc02a695bbaeb198ad867501
  - Removes the TypedArrayFrom abstraction in favour of simpler algorithms
  - (A bit more work is required in implementations which want to avoid allocating the temporary array object created in IterableToArrayLike. In self-hosted environments this may not be an issue, e.g. V8 works quite similar to the proposed changes.)

**Comment by anba** [comment]
WDYT about changing TypedArrayCreate and TypedArraySpeciesCreate to accept an optional _minimumLength_ parameter and use that one instead of the "If _argumentList_ is a List of a single Number, ..." condition in TypedArrayCreate? Will this improve the current situation?

```
<emu-clause id="typedarray-create" aoid="TypedArrayCreate">
  <h1>TypedArrayCreate( _constructor_, _argumentList_ [ , _minimumLength_ ] )</h1>
  <p>The abstract operation TypedArrayCreate with arguments _constructor_ and _argumentList_ is used to specify the creation of a new TypedArray object using a constructor function. The optional argument _minimumLength_ (a positive integer) is used to validate the newly created TypedArray object's array length. TypedArrayCreate performs the following steps:</p>
  <emu-alg>
    1. Let _newTypedArray_ be ? Construct(_constructor_, _argumentList_).
    1. Perform ? ValidateTypedArray(_newTypedArray_).
    1. If _minimumLength_ is present,
      1. Assert: _minimumLength_ is an integer Number &ge; 0.
      1. If the value of _newTypedArray_'s [[ArrayLength]] internal slot < _minimumLength_, throw a *TypeError* exception.
    1. Return _newTypedArray_.
  </emu-alg>
</emu-clause>

<emu-clause id="typedarray-species-create" aoid="TypedArraySpeciesCreate">
  <h1>TypedArraySpeciesCreate( _exemplar_, _argumentList_ [ , _minimumLength_ ] )</h1>
  <p>The abstract operation TypedArraySpeciesCreate with arguments _exemplar_, _argumentList_, and optionally _minimumLength_ is used to specify the creation of a new TypedArray object using a constructor function that is derived from _exemplar_. It performs the following steps:</p>
  <emu-alg>
    1. Assert: _exemplar_ is an Object that has a [[TypedArrayName]] internal slot.
    1. Let _defaultConstructor_ be the intrinsic object listed in column one of <emu-xref href="#table-49"></emu-xref> for the value of _exemplar_'s [[TypedArrayName]] internal slot.
    1. Let _constructor_ be ? SpeciesConstructor(_exemplar_, _defaultConstructor_).
    1. Return ? TypedArrayCreate(_constructor_, _argumentList_). If _minimumLength_ was passed as a parameter then pass its value as the _minimumLength_ optional argument of TypedArrayCreate.
  </emu-alg>
</emu-clause>
```

**Comment by bterlson** [comment]
@anba this change looks great. Really good catch on the mistakenly added recursiveness and I think the refactorings you've made help clarify this text a lot.

I am also a little uncomofrtable with the "List of a single Number" step but I don't feel strongly enough about it that I think we should add an additional parameter (and pass it appropriately for (admittedly few) calls) to avoid it. That said, I leave it up to you if you want to add it.

**Comment by bterlson** [comment]
I also rebased this with latest master. There were some conflicts. If you would like to review this, the commit is here: https://github.com/bterlson/ecma262/commit/eadbf0b646d167b04faea58b504a06e7a1174c28.

**Comment by anba** [comment]
> @anba this change looks great. Really good catch on the mistakenly added recursiveness and I think the refactorings you've made help clarify this text a lot.

Thanks!

> I am also a little uncomofrtable with the "List of a single Number" step but I don't feel strongly enough about it that I think we should add an additional parameter (and pass it appropriately for (admittedly few) calls) to avoid it. 

That's also my opinion on this. So let's just defer this change for now.

> There were some conflicts. If you would like to review this, the commit is here: bterlson@eadbf0b.

Looks good, except a pre-existing bug in my PR (step 4 in sec-typedarray-buffer-byteoffset-length has `</code>%<var>TypedArray` instead of `<code>%<var>TypedArray`).
